docs/Modern-TypeEmission.md

```diff
@@ -0,0 +1,325 @@
+
+<QueryRenderer
+  query={graphql`
+    query ExampleQuery {
+      artist(id: "banksy") {
+        name
+      }
+    }
+  `}
+  render={({ props }: { props?: ExampleQueryResponse }) => {
+    if (props) {
+      return props.artist && <div>{props.artist.name} is great!</div>
+    }
+    return <div>Loading</div>
+  }}
+/>
+```
+<!--TypeScript-->
+```javascript
+/**
+ * export type ExampleQueryResponse = {
+ *   readonly artist?: {
+ *     readonly name: string
+ *   }
+ * }
+ * export type ExampleQuery = {
+ *   readonly response: ExampleQueryResponse
+ * }
+ */
+import { ExampleQuery } from "__generated__/ExampleQuery.graphql"
+
+<QueryRenderer<ExampleQuery>
+  query={graphql`
+    query ExampleQuery {
+      artist(id: "banksy") {
+        name
+      }
+    }
+  `}
+  render={({ props }) => {
+    if (props) {
+      return props.artist && <div>{props.artist.name} is great!</div>
+    }
+    return <div>Loading</div>
+  }}
+/>
+```
+<!--END_DOCUSAURUS_CODE_TABS-->
+
+Similarly, in this example the emitted type-information describes the prop data that the container expects to receive.
+
+<!--DOCUSAURUS_CODE_TABS-->
+<!--Flow-->
+```javascript
+/**
+ * export type ExampleFragment_artist = {|
+ *   +name: string
+ * |}
+ */
+import type { ExampleFragment_artist } from "__generated__/ExampleFragment_artist.graphql"
+
+export const ExampleFragment = createFragmentContainer(
+  (props: { artist: ExampleFragment_artist }) => (
+    <div>About the artist: {props.artist.biography}</div>
+  ),
+  graphql`
+    fragment ExampleFragment_artist on Artist {
+      biography
+    }
+  `
+)
+```
+<!--TypeScript-->
+```javascript
+/**
+ * export type ExampleFragment_artist = {
+ *   readonly name: string
+ * }
+ */
+import { ExampleFragment_artist } from "__generated__/ExampleFragment_artist.graphql"
+
+export const ExampleFragment = createFragmentContainer(
+  (props: { artist: ExampleFragment_artist }) => (
+    <div>About the artist: {props.artist.biography}</div>
+  ),
+  graphql`
+    fragment ExampleFragment_artist on Artist {
+      biography
+    }
+  `
+)
+```
+<!--END_DOCUSAURUS_CODE_TABS-->
+
+### Fragment references
+
+The opaque identifier described in [data-masking] that a child container expects to receive from its parent, which represents the child container’s fragment spread inside the parent’s fragment.
+
+_Please read [this important caveat](#single-artifact-directory) about actually enabling type-safe fragment reference checking._
+
+Consider a component that composes the above fragment container example. In this example, the emitted type-information of the child container receives a unique opaque identifier type, called a fragment reference, which the type-information emitted for the parent’s fragment references in the location where the child’s fragment is spread. Thus ensuring that the child’s fragment is spread into the parent’s fragment _and_ the correct fragment reference is passed to the child container at runtime.
+
```

Comment:
This section is the least clear to me – I think some additional notes on why and how you might use this would help. Maybe even a sentence like:

> Fragment references are how Relay connects different fragment containers together based on which fragments they themselves spread.

Or something like that. It's a tricky concept to try to describe 😅